zebra-network/src/peer/connection.rs

```diff
@@ -369,454 +516,226 @@ where
+                .await;
+
+            if matches!(transition, Transition::AwaitResponse { .. }) {
+                self.request_timer = Some(sleep(constants::REQUEST_TIMEOUT));
             }
-        }
-    }
 
-    /// Marks the peer as having failed with error `e`.
-    fn fail_with<E>(&mut self, e: E)
-    where
-        E: Into<SharedPeerError>,
-    {
-        let e = e.into();
-        debug!(%e,
-               connection_state = ?self.state,
-               client_receiver = ?self.client_rx,
-               "failing peer service with error");
-        // Update the shared error slot
-        let mut guard = self
-            .error_slot
-            .0
-            .lock()
-            .expect("mutex should be unpoisoned");
-        if let Some(original_error) = guard.clone() {
-            // This panic typically happens due to these bugs:
-            // * we mark a connection as failed without using fail_with
-            // * we call fail_with without checking for a failed connection
-            //   state
-            //
-            // See the original bug #1510 and PR #1531, and the later bug #1599
-            // and PR #1600.
-            panic!(
-                "calling fail_with on already-failed connection state: failed connections must stop processing pending requests and responses, then close the connection. state: {:?} original error: {:?} new error: {:?} client receiver: {:?}",
-                self.state,
-                original_error,
-                e,
-                self.client_rx
-            );
-        } else {
-            *guard = Some(e);
-        }
-        // Drop the guard immediately to release the mutex.
-        std::mem::drop(guard);
-
-        // We want to close the client channel and set State::Failed so
-        // that we can flush any pending client requests. However, we may have
-        // an outstanding client request in State::AwaitingResponse, so
-        // we need to deal with it first if it exists.
-        self.client_rx.close();
-        let old_state = std::mem::replace(&mut self.state, State::Failed);
-        if let State::AwaitingResponse { tx, .. } = old_state {
-            // We know the slot has Some(e) because we just set it above,
-            // and the error slot is never unset.
-            let e = self.error_slot.try_get_error().unwrap();
-            let _ = tx.send(Err(e));
+            self.state = match transition.try_into() {
+                Ok(state) => Some(state),
+                Err(e) => {
+                    // while let Some(InProgressClientRequest { tx, span, .. }) =
+                    //     self.client_rx.next().await
+                    // {
+                    //     trace!(
+                    //         parent: &span,
+                    //         "sending an error response to a pending request on a failed connection"
+                    //     );
+                    //     let _ = tx.send(Err(e.clone()));
+                    // }
```

Comment:
Can we just drop all these requests, or do we need to return errors on them?

Reply:
I'm going to look into this today, and see what we might be able to fix here.

Reply:
See 4c75fa27ab7ddc87d052d4ee5adc8ed9e254ceec, where I:
- Add an `ExitClient` transition, used when the internal client channel is closed or dropped, and there are no more pending requests
- Ignore pending requests after an `ExitClient` transition, because they are impossible
- Reject pending requests when the peer has caused an error (the `Exit` and `ExitRequest` transitions)
- Remove `PeerError::ConnectionDropped`, because it is now handled by `ExitClient`. (Which is an internal error, not a peer error.)